packages/gatsby/src/commands/develop.ts

```diff
@@ -1,613 +1,200 @@
-}, 10000)
-
-onExit(() => {
-  telemetry.trackCli(`DEVELOP_STOP`)
-})
-
-type ActivityTracker = any // TODO: Replace this with proper type once reporter is typed
-
-interface IServer {
-  compiler: webpack.Compiler
-  listener: http.Server | https.Server
-  webpackActivity: ActivityTracker
-}
-
-async function startServer(program: IProgram): Promise<IServer> {
-  const indexHTMLActivity = report.phantomActivity(`building index.html`, {})
-  indexHTMLActivity.start()
-  const directory = program.directory
-  const directoryPath = withBasePath(directory)
-  const workerPool = WorkerPool.create()
-  const createIndexHtml = async (activity: ActivityTracker): Promise<void> => {
-    try {
-      await buildHTML({
-        program,
-        stage: Stage.DevelopHTML,
-        pagePaths: [`/`],
-        workerPool,
-        activity,
-      })
-    } catch (err) {
-      if (err.name !== `WebpackError`) {
-        report.panic(err)
-        return
-      }
-      report.panic(
-        report.stripIndent`
-          There was an error compiling the html.js component for the development server.
-
-          See our docs page on debugging HTML builds for help https://gatsby.dev/debug-html
-        `,
-        err
-      )
-    }
+import tmp from "tmp"
+import { spawn } from "child_process"
+import chokidar from "chokidar"
+import resolveCwd from "resolve-cwd"
+import getRandomPort from "get-port"
+import socket from "socket.io"
+import fs from "fs-extra"
+import { createServiceLock } from "gatsby-core-utils/dist/service-lock"
+import { startDevelopProxy } from "../utils/develop-proxy"
+import { IProgram } from "./types"
+
+// Adapted from https://stackoverflow.com/a/16060619
+const requireUncached = (file: string): any => {
+  try {
+    delete require.cache[require.resolve(file)]
+  } catch (e) {
+    return null
   }
 
-  await createIndexHtml(indexHTMLActivity)
-
-  indexHTMLActivity.end()
-
-  // report.stateUpdate(`webpack`, `IN_PROGRESS`)
-
-  const webpackActivity = report.activityTimer(`Building development bundle`, {
-    id: `webpack-develop`,
-  })
-  webpackActivity.start()
-
-  const devConfig = await webpackConfig(
-    program,
-    directory,
-    `develop`,
-    program.port,
-    { parentSpan: webpackActivity.span }
-  )
-
-  const compiler = webpack(devConfig)
+  try {
+    return require(file)
+  } catch (e) {
+    return null
+  }
+}
 
-  /**
-   * Set up the express app.
-   **/
-  const app = express()
-  app.use(telemetry.expressMiddleware(`DEVELOP`))
-  app.use(
-    webpackHotMiddleware(compiler, {
-      log: false,
-      path: `/__webpack_hmr`,
-      heartbeat: 10 * 1000,
-    })
+// Heuristics for gatsby-config.js, as not all changes to it require a full restart to take effect
+const doesConfigChangeRequireRestart = (
+  lastConfig: Record<string, any>,
+  newConfig: Record<string, any>
+): boolean => {
+  // Ignore changes to siteMetadata
+  if (
+    JSON.stringify({ ...lastConfig, siteMetadata: null }) ===
```

Comment:
`config` _may_ be non serializable by `JSON.stringify` (It can contain functions)

I'm unsure if this check will work in that case or throw an Exception? @pieh will know more.

Reply:
This will _not_ throw but won't catch changes to functions in config (for instance, `gatsby-plugin-feed`)

Reply:
This works now